rules/TypeDeclaration/Rector/FunctionLike/AddParamTypeForFunctionLikeWithinCallLikeArgDeclarationRector.php

```diff
@@ -0,0 +1,220 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\TypeDeclaration\Rector\FunctionLike;
+
+use PhpParser\Node;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\CallLike;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\FunctionLike;
+use PhpParser\Node\Identifier;
+use PhpParser\Node\Param;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use Rector\Contract\Rector\ConfigurableRectorInterface;
+use Rector\NodeTypeResolver\TypeComparator\TypeComparator;
+use Rector\Php\PhpVersionProvider;
+use Rector\PHPStanStaticTypeMapper\Enum\TypeKind;
+use Rector\Rector\AbstractRector;
+use Rector\StaticTypeMapper\StaticTypeMapper;
+use Rector\TypeDeclaration\ValueObject\AddParamTypeForFunctionLikeWithinCallLikeArgDeclaration;
+use Rector\ValueObject\PhpVersionFeature;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\ConfiguredCodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+use Webmozart\Assert\Assert;
+
+/**
+ * @see \Rector\Tests\TypeDeclaration\Rector\FunctionLike\AddParamTypeForFunctionLikeWithinCallLikeArgDeclarationRector\AddParamTypeForFunctionLikeWithinCallLikeArgDeclarationRectorTest
+ */
+final class AddParamTypeForFunctionLikeWithinCallLikeArgDeclarationRector extends AbstractRector implements ConfigurableRectorInterface
+{
+    /**
+     * @var AddParamTypeForFunctionLikeWithinCallLikeArgDeclaration[]
+     */
+    private array $addParamTypeForFunctionLikeParamDeclarations = [];
+
+    private bool $hasChanged = false;
+
+    public function __construct(
+        private readonly TypeComparator $typeComparator,
+        private readonly PhpVersionProvider $phpVersionProvider,
+        private readonly StaticTypeMapper $staticTypeMapper,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add param types where needed', [
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+(new SomeClass)->process(function ($parameter) {});
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+(new SomeClass)->process(function (string $parameter) {});
+CODE_SAMPLE
+                ,
+                [
+                    new AddParamTypeForFunctionLikeWithinCallLikeArgDeclaration(
+                        'SomeClass',
+                        'process',
+                        0,
+                        0,
+                        new StringType()
+                    ),
+                ]
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [MethodCall::class, StaticCall::class];
+    }
+
+    /**
+     * @param CallLike $node
+     */
+    public function refactor(Node $node): ?Node
+    {
```

Comment:
`$this->hasChanged` needs to be re-init to false early here as rector rule is shared service:

```suggestion
    {
        $this->hasChanged = false;
```